File: src/offerservice/src/plugins/open-feature/pluginProvider.ts
/* eslint-disable @typescript-eslint/no-unused-vars */
import {
    EvaluationContext,
    JsonValue,
    Logger,
    Provider,
    ResolutionDetails,
} from "@openfeature/server-sdk"
import { FeatureFlagClient } from "../feature-flag/client"
export class PluginProvider implements Provider {
    private featureFlagClient = new FeatureFlagClient()

    metadata = {
        name: "Plugin Provider",
    }

    async resolveBooleanEvaluation(
        flagKey: string,
        defaultValue: boolean,
        context: EvaluationContext
    ): Promise<ResolutionDetails<boolean>> {
        const plugin = await this.featureFlagClient.getFlag(flagKey)

        return {
            value: plugin?.enabled ?? defaultValue,
            variant: flagKey,
        }
    }

    resolveStringEvaluation(
        flagKey: string,
        defaultValue: string,
        context: EvaluationContext
    ): Promise<ResolutionDetails<string>> {
        throw new Error("Method not implemented.")
    }
    resolveNumberEvaluation(
        flagKey: string,
        defaultValue: number,
        context: EvaluationContext
    ): Promise<ResolutionDetails<number>> {
        throw new Error("Method not implemented.")
    }
    resolveObjectEvaluation<T extends JsonValue>(
        flagKey: string,
        defaultValue: T,
        context: EvaluationContext,
        logger: Logger
    ): Promise<ResolutionDetails<T>> {
        throw new Error("Method not implemented.")
    }
}

File: src/broker-service/src/ProblemPatterns/OpenFeature/Providers/PluginProvider.cs
using EasyTrade.BrokerService.ProblemPatterns.OpenFeature.Providers.FeatureFlagService;
using OpenFeature;
using OpenFeature.Model;

namespace EasyTrade.BrokerService.ProblemPatterns.OpenFeature.Providers;

public class PluginProvider(IFeatureFlagServiceConnector flagServiceConnector) : FeatureProvider
{
    public static string Name => "Plugin Provider";
    private readonly IFeatureFlagServiceConnector _flagServiceConnector = flagServiceConnector;

    public override Metadata GetMetadata() => new(Name);

    public override async Task<ResolutionDetails<bool>> ResolveBooleanValueAsync(
        string flagKey,
        bool defaultValue,
        EvaluationContext? context = null,
        CancellationToken cancellationToken = default
    )
    {
        var flag = await _flagServiceConnector.GetFlag(flagKey);
        var value = flag is null ? defaultValue : flag.Enabled;
        return new ResolutionDetails<bool>(flagKey, value);
    }

    public override Task<ResolutionDetails<double>> ResolveDoubleValueAsync(
        string flagKey,
        double defaultValue,
        EvaluationContext? context = null,
        CancellationToken cancellationToken = default
    ) => throw new NotImplementedException();

    public override Task<ResolutionDetails<int>> ResolveIntegerValueAsync(
        string flagKey,
        int defaultValue,
        EvaluationContext? context = null,
        CancellationToken cancellationToken = default
    ) => throw new NotImplementedException();

    public override Task<ResolutionDetails<string>> ResolveStringValueAsync(
        string flagKey,
        string defaultValue,
        EvaluationContext? context = null,
        CancellationToken cancellationToken = default
    ) => throw new NotImplementedException();

    public override Task<ResolutionDetails<Value>> ResolveStructureValueAsync(
        string flagKey,
        Value defaultValue,
        EvaluationContext? context = null,
        CancellationToken cancellationToken = default
    ) => throw new NotImplementedException();
}
